src/Ilios/CoreBundle/Service/CurriculumInventory/ReportRollover.php

```diff
@@ -0,0 +1,176 @@
+     * @param int|null $newYear Academic year override for the rolled-over report.
+     * @return CurriculumInventoryReportInterface The report created during rollover.
+     */
+    public function rollover(
+        CurriculumInventoryReportInterface $report,
+        $newName = null,
+        $newDescription = null,
+        $newYear = null
+    ) {
+        /* @var CurriculumInventoryReportInterface $newReport */
+        $newReport = $this->reportManager->create();
+        $newReport->setStartDate($report->getStartDate());
+        $newReport->setEndDate($report->getEndDate());
+        $newReport->setProgram($report->getProgram());
+        if (isset($newName)) {
+            $newReport->setName($newName);
+        } else {
+            $newReport->setName($report->getName());
+        }
+        if (isset($newDescription)) {
+            $newReport->setDescription($newDescription);
+        } else {
+            $newReport->setDescription($report->getDescription());
+        }
+        if (isset($newYear)) {
+            $newReport->setYear($newYear);
+        } else {
+            $newReport->setYear($report->getYear());
+        }
+        $this->reportManager->update($newReport, false, false);
+
+        $newLevels = [];
+        $levels = $report->getAcademicLevels();
+        foreach ($levels as $level) {
+            /* @var CurriculumInventoryAcademicLevelInterface $newLevel */
+            $newLevel = $this->academicLevelManager->create();
+            $newLevel->setLevel($level->getLevel());
+            $newLevel->setName($level->getName());
+            $newLevel->setDescription($level->getDescription());
+            $newReport->addAcademicLevel($newLevel);
+            $newLevel->setReport($newReport);
+            $this->academicLevelManager->update($newLevel, false, false);
+            $newLevels[$newLevel->getLevel()] = $newLevel;
+        }
+
+        // recursively rollover sequence blocks.
+        $topLevelBlocks = $report
+            ->getSequenceBlocks()
+            ->filter(function (CurriculumInventorySequenceBlockInterface $block) {
+                return is_null($block->getParent());
+            });
+
+        foreach ($topLevelBlocks as $block) {
+            $this->rolloverSequenceBlock($block, $newReport, $newLevels, null);
+        }
+
+        $sequence = $report->getSequence();
+        /* @var  CurriculumInventorySequenceInterface $newSequence */
+        $newSequence = $this->sequenceManager->create();
+        $newSequence->setDescription($sequence->getDescription());
+        $newReport->setSequence($newSequence);
+        $newSequence->setReport($newReport);
+        $this->sequenceManager->update($newSequence, true, false); // flush here.
+
+
+        // generate token after the fact and persist report once more.
+        $newReport->generateToken();
+        $this->reportManager->update($newReport, true, true);
```

Comment:
token generation needs to happen after initial save. then the record must be saved again. this second save is legit.